### PR TITLE
Various improvements, mostly to the bidict example

### PR DESCRIPTION
**Pull request by jab**
Also stub out a "Prior Art" section that lists https://returns.readthedocs.io/en/latest/pages/hkt.html.

### CODE REVIEW

**Comment by nekitdev** [comment]
LGTM! Even though `returns` is slightly hacky imho, it is a good example of HKT demand in python.

**Comment by jab** [comment]
Oh interesting, I haven't looked at returns much beyond just knowing that it exists, would be curious to hear more. (May even be worth discussing the hackiness in the PEP, at least with respect to the HKT implementation?)

Thanks for merging!

**Comment by nekitdev** [comment]
NP. The main issue is that HKTs in `returns` are essentially opt-in (via inheriting from `SupportdKindN`), which makes it similar to only allowing to parametrize `Self` (i.e. functions are unable to utilize HKTs). I could be wrong, though, so it'd be great if someone could confirm this.

**Comment by nekitdev** [comment]
The main *hackiness* goes down to how it looks on the [mypy](https://github.com/python/mypy) plugin side, which is really messy due to the fact HKTs are not native at the moment.

**Comment by jab** [comment]
Ah, in that case any unavoidable hackiness probably isn't fair to discuss in the PEP. But any concerns about the API/UX/etc. (e.g. the opt-in aspect that you mentioned) sounds like fair game.

**Comment by nekitdev** [comment]
I agree! I think we're only going to mention the opt-in aspect and the fact that using HKTs still is going to look slightly unpleasant unless they become native. (like `HKT1[I, T]` instead of `I[T]` all over the place is a bit wordy)

**Comment by jab** [comment]
Ah yeah, putting that in somewhere would be gold! 💯

**Comment by nekitdev** [comment]
I think a very good example I stumbled upon is https://sobolevn.me/2020/10/higher-kinded-types-in-python, in a way how it goes from "simple and logical" to arguably "messy and requiring more thought".

**Comment by nekitdev** [comment]
Well, there even is a section called "hacks", haha :p

**Comment by jab** [comment]
hah just saw that too! Gotta hand it to him for pulling it off at all given the constraints, to say nothing of the great post explaining everything!

Say, in case he's missed the latest on this and on https://github.com/python/typing/issues/548, think it's worth giving him a ping, to make sure he's aware of this PEP which will be mentioning his work?

That also reminds me, there was mention on https://github.com/python/typing/issues/548 of other languages with native HKTs (Scala and Haskell, IIRC), in which case maybe that deserves a brief discussion under "Prior Art" too? (Maybe requires recruiting someone with more familiarity with those languages to help write or review (and ideally there'd be at least one gradually-typed language... maybe Typed Racket has them?), but in return, this could end up informing the spec and implementation in key ways. OTOH, if an initial version of the PEP ends up omitting any mention of other languages with HKTs, there may be an immediate objection until that's added. </$.02!>

**Comment by nekitdev** [comment]
I agree, we should definitely have a section showing other languages, even if that forces us to sprinkle some functional programming here and there, haha. I'd say Haskell is the prime example of HKTs, TBH.

**Comment by nekitdev** [comment]
Also yeah, a ping makes sense. @sobolevn, we're currently working on a PEP introducing HKTs, and any input / suggestions are highly appreciated! :)

**Comment by jab** [comment]
(Yeah, I figured about Haskell. Hopefully we can find an example of a gradually typed language that successfully added HKTs to help make the case. I did a quick search for Typed Racket but so far only found a [thread from 2017](https://groups.google.com/g/racket-users/c/O4vUaEPYez4/m/ULpZRt10BAAJ), and based on skimming, HKTs weren't supported at the time but they were hoping to add them in the future.

File: pep-9999.rst [hunk 1]
```diff
@@ -67,26 +67,45 @@ HKTs have long been a requested feature, for quite some use cases [#typing-548]_
 
 .. code:: python
 
+    from collections import Mapping, MutableMapping
+
```
Comment (jab) [inline pep-9999.rst:71]: Should have been `collections.abc` 🤦‍♂️

**Comment by sobolevn** [comment]
Oh, awesome! I will take a look today!

**Comment by sobolevn** [comment]
My feedback:
1. I really recommend to start a discussion in `#typing-sig` before actually going all-in with this PEP. This is something a lot of people will have their opinion on. Maybe even speaking at typing meetup is required. Big features are discussed for a long time 🙂 And this discussion can really improve the chances of getting accepted
2. The most important part in each PEP is "Motivation". This is going to be quite hard. Because we can use either syntetic examples or function programming ones. Both are not quite convincing. Can you find builtin functions / methods that need this? Do some popular libraries need this (except `bidict`)? 
3. I would also not recommend to mix concepts here. I think it is safe to leave new type vars syntax out for now. This PEP is quite new, let's be as simple as possible.
4. `type Unary[T, U] = Callable[[T], U]` seems unneeded, `Callable[[T], U]` is good enough

I am very interested in getting this into Python, if you need any kind of my help - please, let me know.

**Comment by AlexWaygood** [comment]
> 2\. Can you find builtin functions / methods that need this?

I still haven't read the draft PEP yet, but when I have time (maybe at the weekend), I'm hoping to put together a list of problems we have at typeshed that would be solved by having HKTs. The classic one is `dict.fromkeys()`.

**Comment by nekitdev** [comment]
Regarding "Motivation" section, I definitely agree. What do you mean by new type syntax? I guess we need to resort to non-PEP 695, `I = TypeVar("I", ...)` syntax?
Then we need to figure out how exactly they're going to be represented.

**Comment by AlexWaygood** [comment]
> @AlexWaygood are you working on this as well? Or is just a file template? 🙂

Haven't put in any work yet, but I'm watching the repo and I'm interested in helping out as soon as I have the time!

**Comment by nekitdev** [comment]
> > 2\. Can you find builtin functions / methods that need this?
> I still haven't read the draft PEP yet, but when I have time (maybe at the weekend), I'm hoping to put together a list of problems we have at typeshed that would be solved by having HKTs. The classic one is `dict.fromkeys()`.

Huh, great catch! Haven't thought about this one.

**Comment by nekitdev** [comment]
I'm proposing some argument to `TypeVar`, like `args`, which contains the count of `TypeVar`s the type variable can be parametrized with; it would be `0` by default for compatibility reasons.
Any other suggestions on how we could implement it?

**Comment by nekitdev** [comment]
I don't think we're going to allow `ParamSpec`s or `TypeVarTuple`s into the parametrization, though, as it makes things even more complex.

**Comment by nekitdev** [comment]
Setting `args` to `None` could be used to signal `TypeVarTuple` usage, and `ParamSpec`s can be treated as `TypeVar`s.

**Comment by hmc-cs-mdrissi** [comment]
I’ll try to find some examples you can use this weekend from numerical libraries. I have some code that could use HKT involving tensorflow/numpy/beam ecosystem and have been leaving comments of Any due to lack of them. 

Data science ecosystem itself is a good place to find some functional code that could benefit. Map reduce style libraries like beam/spark are functional enough that some of the core types ideally should use HKT.

**Comment by nekitdev** [comment]
I really like the `kind` argument idea. Though, I'm not sure if it makes sense for non-bounded type variables, i.e. it implies "any type that can be parametrized like so" or something like that.

**Comment by nekitdev** [comment]
Regarding the `kind` argument, I feel that something like

```python
Q = TypeVar("Q", bound=Hashable)  # `kind` is `()` by default

I = TypeVar("I", bound=Iterable[T], kind=T)

M = TypeVar("M", bound=Mapping[K, V], kind=(K, V))
```

can work really well. Any opinions?

**Comment by sobolevn** [comment]
Do we need to list all typevars? Or can we just go with `kinded=True`?

**Comment by nekitdev** [comment]
We could do that, yeah, but listing all `TypeVar`s works better in a sense one can check that the amount of arguments is correct.

Also, `kind=...` works in the same way as `K[...]`, so that's one of the reasons I tend towards it.

**Comment by hmc-cs-mdrissi** [comment]
I did expect to use kind without specifying bound.

F = TypeVar(“F”, kind=T1) means any type that takes one type argument. My main experience with hkt comes from haskell and you are not required to have a bound when using an hkt. As an example which bound should I use here,

```python
class Functor(Protocol[F])
  def fmap(self, f: Callable[[A], B], x: F[A]) -> F[B]:
  …
```

A protocol hkt may make sense to be unbounded like this.

If you specify bound then I think there’s no real need to specify type arguments again besides explicitness. I’m neutral to specifying them again. If we have implicit rule though main thing is be cautious of edge cases and not changing existing meaning for working code today.

**Comment by nekitdev** [comment]
Manually specifying kind arguments makes sense since HKTs should imo be opt-in, basically.
